Sources/GRPCHTTP2Core/Client/Connection/LoadBalancers/PickFirstLoadBalancer.swift

```diff
@@ -0,0 +1,530 @@
+  func pickSubchannel() -> Subchannel? {
+    let onPickSubchannel = self.state.withLockedValue { $0.pickSubchannel() }
+    switch onPickSubchannel {
+    case .picked(let subchannel):
+      return subchannel
+    case .notAvailable(let subchannel):
+      subchannel?.connect()
+      return nil
+    }
+  }
+}
+
+@available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *)
+extension PickFirstLoadBalancer {
+  private func handleUpdateEndpoint(_ endpoint: Endpoint, in group: inout DiscardingTaskGroup) {
+    if endpoint.addresses.isEmpty { return }
+
+    let onUpdate = self.state.withLockedValue { state in
+      state.updateEndpoint(endpoint) { endpoint, id in
+        Subchannel(
+          endpoint: endpoint,
+          id: id,
+          connector: self.connector,
+          backoff: self.backoff,
+          defaultCompression: self.defaultCompression,
+          enabledCompression: self.enabledCompression
+        )
+      }
+    }
+
+    switch onUpdate {
+    case .connect(let newSubchannel, close: let oldSubchannel):
+      self.runSubchannel(newSubchannel, in: &group)
+      oldSubchannel?.close()
+
+    case .none:
+      ()
+    }
+  }
+
+  private func runSubchannel(
+    _ subchannel: Subchannel,
+    in group: inout DiscardingTaskGroup
+  ) {
+    // Start running it and tell it to connect.
+    subchannel.connect()
+    group.addTask {
+      await subchannel.run()
+    }
+
+    group.addTask {
+      for await event in subchannel.events {
+        switch event {
+        case .connectivityStateChanged(let state):
+          self.handleSubchannelConnectivityStateChange(state, id: subchannel.id)
+        case .goingAway:
+          ()  // self.handleSubchannelGoingAway(key: key)
+        case .requiresNameResolution:
+          self.event.continuation.yield(.requiresNameResolution)
+        }
+      }
+    }
+  }
+
+  private func handleSubchannelConnectivityStateChange(
+    _ connectivityState: ConnectivityState,
+    id: SubchannelID
+  ) {
+    let onUpdateState = self.state.withLockedValue {
+      $0.updateSubchannelConnectivityState(connectivityState, id: id)
+    }
+
+    switch onUpdateState {
+    case .close(let subchannel):
+      subchannel.close()
+    case .closeAndPublishStateChange(let subchannel, let connectivityState):
+      subchannel.close()
+      self.event.continuation.yield(.connectivityStateChanged(connectivityState))
+    case .publishStateChange(let connectivityState):
+      self.event.continuation.yield(.connectivityStateChanged(connectivityState))
+    case .closed:
+      self.event.continuation.finish()
+      self.input.continuation.finish()
+    case .none:
+      ()
+    }
+  }
+
+  private func handleCloseInput() {
+    let onClose = self.state.withLockedValue { $0.close() }
+    switch onClose {
+    case .closeSubchannels(let subchannel1, let subchannel2):
+      // TODO: when does this finish?
+      self.event.continuation.yield(.connectivityStateChanged(.shutdown))
```

Comment:
Should this be finished after this yield? Are we supposed to get more events after we've emitted shutdown?

Reply:
Nope, this was a leftover TODO before I added `.closed` which is where it should be done. The reason we wait is that that `.shutdown` is emitted when the LB begins shutting down. We still need to wait for the existing subchannels to shutdown.